Add betabinomial logpmf/pmf and tests

Dpananos opened the pull request:
Similar to  #5608, this PR adds logpmf/pmf for betabinomial in jax.scipy.stats

Tests included.

jakevdp commented on tests/scipy_stats_test.py:
```diff
@@ -420,6 +420,23 @@ def args_maker():
 
+  @genNamedParametersNArgs(5)
+  def testBetaBinomLogPmf(self, shapes, dtypes):
+    rng = jtu.rand_positive(self.rng())
```
Would it be worth testing negative inputs to make sure they correctly return infinity?

Dpananos replied:
Ah, good catch.  Are you thinking maybe `rng = jtu.rand_default(self.rng())`?  That is what is used to test the bernoulli and Poisson.

jakevdp replied:
Yeah something like that – particularly if the return value in this range is something explicit like `inf`

majnemer commented on jax/_src/scipy/stats/betabinom.py:
```diff
@@ -0,0 +1,50 @@
+    
+    combiln = lax.neg(
+        lax.add(
+            lax.log(lax.add(n, one)),
```
Should `lax.log(lax.add(n, one))` be replaced with `lax.log1p(n)`?

Dpananos replied:
Yes!  Still learning some of the `lax` functions, so if you see more of that let me know.

Dpananos commented:
I'm runningn into a small problem with this PR re: testing as per @jakevdp 's suggestion.  When I use `jtu.rand_default` in the test, the parameters generated look something like the following:

```python

k = [1., -0.38,  1.36, -0.84]
n = [[ 2.0,  4.23,  0.7,  -1.07],
     [ 0.01,  0.7,   2.08,  2.5 ],
     [-0.92,  0.74, -3.81, -0.54]]

a = [[ 2.77, -0.7,   3.07, -4.9 ],
     [-2.61, -3.69, -3.24, -0.5 ],
     [ 1.78,  5.4,  -1.03,  1.2 ]]

b = [[-0.42],
     [-0.68],
     [-0.04]]

loc = [[0.],
       [ 3.06],
       [-4.31]]
```

If I compare my implementation of the logpmf (without the `where` in the return call), I get something like...

```
import jax.numpy as jnp
from jax import lax
import jax.scipy as lsp

import scipy as osp
import numpy as np

k, n, a, b, loc = map(jnp.array, [k, n, a, b, loc])
lsp.stats.betabinom.logpmf(k=k, n=n, a=a, b=b, loc=loc)

>>>DeviceArray([[ -1.2188828,        -inf,   0.3208853,         nan],
                             [ -7.949671 , -13.266653 , -10.89188  ,  -6.595767 ],
                             [ -1.5104122,  -1.5108628,         nan,  -3.7848833]],            dtype=float32)
```

Comparing this directly with scipy's [implementation of the logpmf](https://github.com/scipy/scipy/blob/v1.6.0/scipy/stats/_discrete_distns.py#L208)

```python
k = np.floor(k) - loc
combiln = -np.log(n + 1) - osp.special.betaln(n - k + 1, k + 1)
combiln + osp.special.betaln(k + a, n - k + b) - osp.special.betaln(a, b)
>>>array([[ -1.2188818,        -inf,   0.3208878,         nan],
                 [ -7.949671 , -13.266655 , -10.891884 ,  -6.595765 ],
                 [ -1.5104126,  -1.5108645,         nan,  -3.784882 ]],
                dtype=float32)
```

Nice, looks similar.  However, calling scipy's betabinom logpmf directly yeilds `nan`

```python
osp.stats.betabinom.logpmf(k=k, n=n, a=a, b=b, loc=loc)
>>>array([[nan, nan, nan, nan],
                 [nan, nan, nan, nan],
                 [nan, nan, nan, nan]])
```

Which is problematic because this is what is compared in the tests.

```
Running tests under Python 3.8.5: /Users/demetri/opt/miniconda3/envs/jax/bin/python
[ RUN      ] LaxBackedScipyStatsTests.testBetaBinomLogPmf_float32[4]_float32[3,4]_float32[3,4]_float32[3,1]_float32[3,1]
W0204 14:08:25.821909 4461870528 xla_bridge.py:135] No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[  FAILED  ] LaxBackedScipyStatsTests.testBetaBinomLogPmf_float32[4]_float32[3,4]_float32[3,4]_float32[3,1]_float32[3,1]
======================================================================
FAIL: testBetaBinomLogPmf_float32[4]_float32[3,4]_float32[3,4]_float32[3,1]_float32[3,1] (__main__.LaxBackedScipyStatsTests)
LaxBackedScipyStatsTests.testBetaBinomLogPmf_float32[4]_float32[3,4]_float32[3,4]_float32[3,1]_float32[3,1]
testBetaBinomLogPmf_float32[4]_float32[3,4]_float32[3,4]_float32[3,1]_float32[3,1](shapes=((4,), (3, 4), (3, 4), (3, 1), (3, 1)), dtypes=(<class 'numpy.float32'>, <class 'numpy.float32'>, <class 'numpy.float32'>, <class 'numpy.float32'>, <class 'numpy.float32'>))
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/demetri/opt/miniconda3/envs/jax/lib/python3.8/site-packages/absl_py-0.11.0-py3.8.egg/absl/testing/parameterized.py", line 282, in bound_param_test
    return test_method(self, **testcase_params)
  File "tests/scipy_stats_test.py", line 433, in testBetaBinomLogPmf
    self._CheckAgainstNumpy(scipy_fun, lax_fun, args_maker, check_dtypes=False,
  File "/Users/demetri/gitrepos/jax/jax/test_util.py", line 915, in _CheckAgainstNumpy
    self.assertAllClose(numpy_ans, lax_ans, check_dtypes=check_dtypes,
  File "/Users/demetri/gitrepos/jax/jax/test_util.py", line 847, in assertAllClose
    self.assertArraysAllClose(x, y, check_dtypes=False, atol=atol, rtol=rtol)
  File "/Users/demetri/gitrepos/jax/jax/test_util.py", line 814, in assertArraysAllClose
    _assert_numpy_allclose(x, y, atol=atol, rtol=rtol)
  File "/Users/demetri/gitrepos/jax/jax/test_util.py", line 133, in _assert_numpy_allclose
    np.testing.assert_allclose(a, b, **kw)
  File "/Users/demetri/opt/miniconda3/envs/jax/lib/python3.8/site-packages/numpy-1.19.5-py3.8-macosx-10.9-x86_64.egg/numpy/testing/_private/utils.py", line 1527, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/Users/demetri/opt/miniconda3/envs/jax/lib/python3.8/site-packages/numpy-1.19.5-py3.8-macosx-10.9-x86_64.egg/numpy/testing/_private/utils.py", line 764, in assert_array_compare
    flagged = func_assert_same_pos(x, y, func=isnan, hasval='nan')
  File "/Users/demetri/opt/miniconda3/envs/jax/lib/python3.8/site-packages/numpy-1.19.5-py3.8-macosx-10.9-x86_64.egg/numpy/testing/_private/utils.py", line 740, in func_assert_same_pos
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=0.0005, atol=0.0005

x and y nan location mismatch:
 x: array([[nan, nan, nan, nan],
       [nan, nan, nan, nan],
       [nan, nan, nan, nan]])
 y: array([[ -2.0639  ,   4.959385,  -4.269639,        nan],
       [-11.134994, -13.355612, -10.916795,  -6.697731],
       [ -1.443631,  -1.582576,        nan,  -3.863479]], dtype=float32)
```

It looks like scipy [does a check to see if [all of `n`, `a`, and `b`](https://github.com/scipy/scipy/blob/v1.6.0/scipy/stats/_discrete_distns.py#L205) are positive.  If not, I think it returns nan.  This might be to prevent singularities in the `betaln` call since the beta function is defined for non negative integers.

So, what should I do in this case?  Should I implement the same sort of argcheck in the logpmf?  If I don't the tests won't pass with `jtu.rand_default`.

Dpananos commented:
I'll also add that the tests pass if I do as other tests do and clip some of the variables.

```
    def args_maker():
      k, n, a, b, loc = map(rng, shapes, dtypes)
      k = np.floor(k)
      a = np.clip(a, a_min = 0.1, a_max = None)
      b = np.clip(b, a_min = 0.1, a_max = None)
      n = np.ceil(n)
      loc = np.floor(loc)
      return [k, n, a, b, loc]
```

jakevdp commented:
Hmm... I don't think scipy's behavior here is necessarily something we want to emulate here.

I think it's fine to only test positive values against scipy and add a separate test that explicitly checks that out-of-range inputs lead to inf or nan where appropriate. What do you think?

Dpananos commented:
Yep, I agree.  Let me see if I can write something up for this then.

Dpananos commented:
I don't see any other examples of testing distributions to see if poor parameterizations result in `inf` or `nan`.  For instance, the `testBernoulliLogPmf` test forces `p` to be between 0 and 1 through application of `expit`.  Additionally, `mu` in `testPoissonLogPmf` is forced to be positive as well.  

If we want to test that the distributions fail in predictable ways:

a) What would be the best way of doing that?  Should I just do something like `self.assertAlClose( inf, lax_fun(k,n,a,b,loc)` (there doesn't seem to be an `assertAllEqual`).

b) Should other distributions (i.e. all bounded distributions like the poisson and bernoulli) be checked in this way?

jakevdp commented:
You're right that we don't have great test coverage for out-of-bound inputs... if you want to ignore that in this case, it's probably OK.

If you do want to tackle it, I'd probably use `self.assertArraysEqual()` (which is a thin wrapper around `np.testing.assert_array_equal`).

Adding such tests to other distributions would be worthwhile as well; it might be worth using something like the `op_record` approach in `lax_numpy_test.py` in order to have a single test of invalid values for multiple distributions. But that doesn't have to be part of this PR if you don't want to go down that route!

Dpananos commented:
OK, this is at a place where I am happy with it.  I'm going to open another issue re: some differences with original scipy and lax scipy.

jakevdp reviewed (approved):
Looks great. It appears that a few test tolerances need to be adjusted, but once tests pass we can get this merged!

Also, please squash all changes into a single commit once you're finished. Thanks!

jakevdp commented:
I'm going to run some additional tests on GPU & TPU backends – I'll let you know if those uncover any issues.

Dpananos commented:
Sweet. Once you run the GPU/TPU checks then I will squash.  Just gimme the thumbs up when ready

Dpananos committed:
Add betabinomial logpmf/pmf and tests

Squash all changes to single commit.  Add betabinom

Add tests for betabinom. nan where undefefined

squash

jakevdp commented:
Everything looks good! We'll get this merged.

jakevdp commented:
Thanks for the contribution! Just FYI I ended up having to do a minor patch to make the new function work with older scipy, which is currently not tested via github CI (I have a PR at #5654 to rectify this).

Dpananos commented:
Cool.  Is there anything I should know for future PRs?

jakevdp commented:
Just that we need compatibility back to scipy 1.2... `betabinom` was added in 1.4. The updated test in #5654 should catch that for future PRs.
